typescript-ignorant/TypeScriptParser.g4

```diff
@@ -0,0 +1,838 @@
+    ;
+
+tupleElementTypes
+    : type_ (',' type_)*
+    ;
+
+functionType
+    : typeParameters? '(' parameterList? ')' '=>' type_
+    ;
+
+constructorType
+    : 'new' typeParameters? '(' parameterList? ')' '=>' type_
+    ;
+
+typeQuery
+    : Typeof typeQueryExpression
+    ;
+
+typeQueryExpression
+    : Identifier
+    | (identifierName '.')+ identifierName
+    ;
+
+propertySignatur
+    : ReadOnly? propertyName '?'? typeAnnotation? ('=>' type_)?
+    ;
+
+typeAnnotation
+    : ':' type_
+    ;
+
+callSignature
+    : typeParameters? '(' parameterList? ')' typeAnnotation?
+    ;
+
+parameterList
+    : restParameter
+    | predefinedType (',' predefinedType)*
+    | optionalParameterList (',' restParameter)?
+    | requiredParameterList (',' (optionalParameterList (',' restParameter)? | restParameter))?
+    ;
+
+requiredParameterList
+    : requiredParameter (',' requiredParameter)*
+    ;
+
+requiredParameter
+    : decoratorList? accessibilityModifier? identifierOrPattern typeAnnotation?
+    ;
+
+accessibilityModifier
+    : Public
+    | Private
+    | Protected
+    ;
+
+identifierOrPattern
+    : identifierName
+    | bindingPattern
+    ;
+
+optionalParameterList
+    : optionalParameter (',' optionalParameter)*
+    ;
+
+optionalParameter
+    : decoratorList? ( accessibilityModifier? identifierOrPattern ('?' typeAnnotation? | typeAnnotation? initializer))
+    ;
+
+restParameter
+    : '...' singleExpression
+    ;
+
+constructSignature
+    : 'new' typeParameters? '(' parameterList? ')' typeAnnotation?
+    ;
+
+indexSignature
+    : '[' Identifier ':' (Number|String) ']' typeAnnotation
+    ;
+
+methodSignature
+    : propertyName '?'? callSignature
+    ;
+
+typeAliasDeclaration
+    : Declare? 'type' Identifier typeParameters? '=' type_ SemiColon
+    ;
+
+constructorDeclaration
+    : accessibilityModifier? Constructor '(' formalParameterList? ')' ( ('{' functionBody '}') | SemiColon)?
+    ;
+
+mappedType
+    : ReadOnly? '[' Identifier In primaryType ']' '?'? typeAnnotation
+    ;
+
+// A.5 Interface
+
+interfaceDeclaration
+    : Declare? Interface Identifier typeParameters? interfaceExtendsClause? objectType SemiColon?
+    ;
+
+interfaceExtendsClause
+    : Extends classOrInterfaceTypeList
+    ;
+
+classOrInterfaceTypeList
+    : typeReference (',' typeReference)*
+    ;
+
+// A.7 Interface
+
+enumDeclaration
+    : Declare? Const? Enum Identifier '{' enumBody? '}'
+    ;
+
+enumBody
+    : enumMemberList ','?
+    ;
+
+enumMemberList
+    : enumMember (',' enumMember)*
+    ;
+
+enumMember
+    : propertyName ('=' singleExpression)?
+    ;
+
+// A.8 Namespaces
+
+namespaceDeclaration
+    : Declare? Namespace namespaceName '{' statementList '}'
+    ;
+
+namespaceName
+    : Identifier ('.' Identifier)*?
+    ;
+
+// Modules look just like namespaces to me
+
+moduleDeclaration
+    : Declare? Module moduleName '{' statementList '}'
+    ;
+
+moduleName
+    : Identifier ('.' Identifier)*?
```

Comment:
If `moduleName` rule equals `namespaceName` maybe merge them and use the single rule something like `dottedId`?